Add ability to compile code from stdin

am11 opened the pull request:
csc, vbc, and csi are now able to compile code from standard input. Plumbing for vbi is also done in this PR, however, there were some unimplemented methods in master, which are required to make the vib project itself functional.

When input is redirected from standard input stream and argument `-` is present, the Console.In is considered as a stream containing source and passed to the compiler; since it is a non-seekable stream, some adjustments has been made in assertions.

Error case: when `-` is present and input stream is not redirected, compiler shall issue a new error `CS8782`.

Closes #15802

am11 commented on src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb:
```diff
@@ -605,6 +607,110 @@ SRC.VB(1) : error BC30037: Character is not valid.
+            If Console.IsInputRedirected Then
+                ' [applicable to both Windows and Unix]
+                ' if our parent (xunit) process itself has input redirected, we cannot test this
+                ' error case because our child process will inherit it and and we cannot achive what
```
```suggestion
                ' error case because our child process will inherit it and we cannot achive what
```

am11 commented on src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:
```diff
@@ -5635,6 +5661,113 @@ public void Utf8Output()
+            {
+                // [applicable to both Windows and Unix]
+                // if our parent (xunit) process itself has input redirected, we cannot test this
+                // error case because our child process will inherit it and and we cannot achive what
```
```suggestion
                // error case because our child process will inherit it and we cannot achive what
```

cston commented on src/Compilers/Core/CodeAnalysisTest/Text/StringTextTest.cs:
```diff
@@ -72,7 +72,6 @@ public void FromStream_Errors()
         {
             Assert.Throws<ArgumentNullException>(() => SourceText.From((Stream)null, Encoding.UTF8));
             Assert.Throws<ArgumentException>(() => SourceText.From(new TestStream(canRead: false, canSeek: true), Encoding.UTF8));
-            Assert.Throws<ArgumentException>(() => SourceText.From(new TestStream(canRead: true, canSeek: false), Encoding.UTF8));
```
>SourceText.From(new TestStream(canRead: true, canSeek: false), Encoding.UTF8) [](start = 51, length = 77)

We should keep this test case, even if it no longer throws an exception.

cston commented on src/Compilers/Core/Portable/CommandLine/CommandLineSourceFile.cs:
```diff
@@ -12,15 +12,13 @@ namespace Microsoft.CodeAnalysis
-        private readonly string _path;
-        private readonly bool _isScript;
-
-        public CommandLineSourceFile(string path, bool isScript)
```
>public CommandLineSourceFile(string path, bool isScript) [](start = 8, length = 56)

We should add a constructor rather than replacing the existing public constructor.

cston commented on src/Compilers/Core/Portable/CommandLine/CommandLineSourceFile.cs:
```diff
@@ -29,17 +27,16 @@ public CommandLineSourceFile(string path, bool isScript)
+        public string Path { get; private set; }
+
+        /// <summary>
+        ///     True if the input has been redirected from the standard input stream.
```
>     [](start = 12, length = 4)

Extra spaces.

cston commented on src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:
```diff
@@ -249,8 +249,15 @@ internal SourceText TryReadFileContent(CommandLineSourceFile file, IList<Diagnos
-                using (var data = OpenFileForReadWithSmallBufferOptimization(filePath))
+                if (file.IsInputRedirected)
                 {
+                    using var data = Console.OpenStandardInput(bufferSize: 1);
```
>bufferSize: 1 [](start = 63, length = 13)

Just curious: Why set an explicit `bufferSize`?

am11 replied:
I used the same buffer size as in case of FileStream per this comment:
https://github.com/dotnet/roslyn/blob/e704ca635bd6de70a0250e34c4567c7a28fa9f6d/src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs#L342-L347

cston replied:
Does that optimization apply to streams where `Seek()` is not supported?

---
In reply to: [373165431](https://github.com/dotnet/roslyn/pull/41166#discussion_r373165431) [](ancestors = 373165431)

am11 replied:
Good point, and https://github.com/dotnet/runtime/blob/b3fa13f/src/libraries/System.Console/src/System/Console.cs#L371-L379 says it all. So I will remove the bufferSize. 😅

cston commented on src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:
```diff
@@ -249,8 +249,15 @@ internal SourceText TryReadFileContent(CommandLineSourceFile file, IList<Diagnos
                 {
+                    using var data = Console.OpenStandardInput(bufferSize: 1);
+                    normalizedFilePath = filePath;
+                    return EncodedStringText.Create(data, Arguments.Encoding, Arguments.ChecksumAlgorithm, canBeEmbedded: EmbeddedSourcePaths.Contains(file.Path));
```
>EmbeddedSourcePaths.Contains(file.Path) [](start = 122, length = 39)

Is this ever `true`? If not, consider replacing with `false`.

am11 replied:
It can be true, as we can compile N files and source for one of them comes from stdin, (which is referred as `stdin.cs`).

cston commented on src/Compilers/Core/Portable/Text/SourceText.cs:
```diff
@@ -280,14 +283,18 @@ private static string Decode(Stream stream, Encoding encoding, out Encoding actu
         {
             RoslynDebug.Assert(stream != null);
             RoslynDebug.Assert(encoding != null);
+            int length = 4096;
```
>length [](start = 16, length = 6)

Consider creating a local called `bufferSize` instead since this does not represent the length of the buffer in the `!CanSeek` case. Then `int length;` and `bufferSize = Math.Min(4096, length);` can be moved inside the `if` block.

cston replied:
>4096 [](start = 25, length = 4)

Perhaps add a `const int minBufferSize = 4096;` since the constant value is used a couple of times.

cston commented on src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs:
```diff
@@ -224,6 +224,19 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string bas
+                            {
+                                if (Console.IsInputRedirected)
+                                {
+                                    sourceFiles.Add(ToCommandLineSourceFile("stdin" + ScriptFileExtension, isInputRedirected: true));
```
>"stdin" + ScriptFileExtension [](start = 76, length = 29)

Is it necessary to give it a different name and add an extension? Or could we just use `new CommandLineSourceFile("-", isScriptFile: true, isInputRedirected: true)` instead?

am11 replied:
It is not necessary and I do not feel strongly about this choice. I just felt it would be more readable in Error List / logs. Also, if we do not provide the output file name, it creates an executable called `stdin.exe` (or `stdin` in case of Unix).

However, on a downside, chances of name collisions are higher for former in `stdin.cs` vs. `-`. In fact, I was trying to create an AppDomain (for net472) in VB's test, and it gave me error that there already is an assembly loaded with the name `stdin.dll` (wasn't the case in an identical C# test), therefore i gave output name in all test cases (but later removed AppDomain altogether as it wasn't supported in netcore).

I will change them to `-`

cston commented on src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:
```diff
@@ -5635,6 +5661,113 @@ public void Utf8Output()
+            assembly.ManifestModule.GetPEKind(out PortableExecutableKinds peKind, out _);
+
+            Assert.Equal(PortableExecutableKinds.ILOnly, peKind);
+        }
```
>} [](start = 8, length = 1)

Please test with duplicate `-` option: `"csc - /out:name.exe -".

cston commented on src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:
```diff
@@ -36,12 +36,38 @@ namespace Microsoft.CodeAnalysis.CSharp.CommandLine.UnitTests
+            s_DotnetCscRun = $@"""{dotnetExe}"" exec --runtimeconfig ""{Path.Combine(Path.GetDirectoryName(cscDllPath), "csc.runtimeconfig.json")}""";
+            s_CSharpScriptExecutable = s_CSharpCompilerExecutable.Replace("csc.dll", Path.Combine("csi", "csi.dll"));
+#else
+            s_CSharpScriptExecutable = s_CSharpCompilerExecutable.Replace("csc.exe", Path.Combine("csi", "csi.exe"));
```
>s_CSharpCompilerExecutable [](start = 39, length = 26)

Is this value `null`?

cston commented on src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:
```diff
@@ -5635,6 +5661,113 @@ public void Utf8Output()
+            Assert.False(result.ContainsErrors, $"Compilation error(s) occurred: {result.Output} {result.Errors}");
+
+            string output = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ?
+                ProcessUtilities.RunAndGetOutput("cmd.exe", $@"/C ""{s_DotnetCscRun} stdin.exe""", expectedRetCode: 0) :
```
>stdin.exe [](start = 85, length = 9)

Where is this file created?

am11 replied:
Current working directory of test: `artifacts\bin\Microsoft.CodeAnalysis.CSharp.CommandLine.UnitTests\{Configuration}\{TFM}`.

cston replied:
Please consider creating a temp directory for each test. Something like:
```C#
var path = Temp.CreateDirectory().Path;
...
ProcessUtilities.RunAndGetOutput("cmd.exe", ..., startFolder: path);

```

---
In reply to: [373214597](https://github.com/dotnet/roslyn/pull/41166#discussion_r373214597) [](ancestors = 373214597)

cston commented on src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:
```diff
@@ -5635,6 +5661,113 @@ public void Utf8Output()
+                // error case because our child process will inherit it and we cannot achive what
+                // we are aiming for: isatty(0):true and thereby Console.IsInputerRedirected:false in
+                // child. running this case will make StreamReader to hang (waiting for input, that
+                // we do not propograte: parent.In->child.In).
```
>propograte [](start = 29, length = 10)

Typo

cston commented:
@am11, please avoid force-push to allowing reviewing changes between commits.

am11 commented:
@cston, sorry about that, i will keep that in mind.
<sub>(btw, GitHub and AzDevOps these days do show the delta between the diffs, regardless of force=1)</sub>

I addressed the remaining feedback, could you please give it another go. 8-)

cston commented on src/Compilers/Core/CodeAnalysisTest/Text/StringTextTest.cs:
```diff
@@ -72,7 +72,7 @@ public void FromStream_Errors()
             Assert.Throws<ArgumentNullException>(() => SourceText.From((Stream)null, Encoding.UTF8));
             Assert.Throws<ArgumentException>(() => SourceText.From(new TestStream(canRead: false, canSeek: true), Encoding.UTF8));
-            Assert.Throws<ArgumentException>(() => SourceText.From(new TestStream(canRead: true, canSeek: false), Encoding.UTF8));
+            Assert.Throws<NotImplementedException>(() => SourceText.From(new TestStream(canRead: true, canSeek: false), Encoding.UTF8));
```
>NotImplementedException [](start = 26, length = 23)

Where are we throwing `NotImplementedException`? If it's from `TestStream`, does that class need to be updated?

am11 replied:
This is no backing stream in this case, so TestStream throws from: https://github.com/dotnet/roslyn/blob/c250d31757c0815c45f344d60c27e50a5e50960e/src/Test/Utilities/Portable/Mocks/TestStream.cs#L102-L109

which seems correct.

cston commented on src/Compilers/Core/Portable/Text/SourceText.cs:
```diff
@@ -280,22 +283,28 @@ private static string Decode(Stream stream, Encoding encoding, out Encoding actu
+                    actualEncoding = encoding;
+                    return string.Empty;
+                }
+                bufferSize = Math.Min(4096, length);
```
>4096 [](start = 38, length = 4)

`minBufferSize`

cston commented on src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:
```diff
@@ -5635,6 +5661,146 @@ public void Utf8Output()
+        }
+
+        [Fact]
+        public void CscCompile_WithRedirectedInputIndicatorAndStandardInputNotRedirected_ThrowsCS8782()
```
>Throws [](start = 89, length = 6)

Perhaps "Reports" or "Error" instead.

cston commented on src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:
```diff
@@ -5635,6 +5661,146 @@ public void Utf8Output()
+                // error case because our child process will inherit it and we cannot achive what
+                // we are aiming for: isatty(0):true and thereby Console.IsInputerRedirected:false in
+                // child. running this case will make StreamReader to hang (waiting for input, that
+                // we do not propagrate: parent.In->child.In).
```
>propagrate [](start = 29, length = 10)

Typo

cston commented on src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:
```diff
@@ -5635,6 +5661,146 @@ public void Utf8Output()
+            {
+                // [applicable to both Windows and Unix]
+                // if our parent (xunit) process itself has input redirected, we cannot test this
+                // error case because our child process will inherit it and we cannot achive what
```
>achive [](start = 86, length = 6)

Typo

cston commented on src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:
```diff
@@ -5635,6 +5661,146 @@ public void Utf8Output()
+            var assembly = Assembly.LoadFrom(Path.Combine(tempDir, name));
+            assembly.ManifestModule.GetPEKind(out PortableExecutableKinds peKind, out _);
+
+            Assert.Equal(PortableExecutableKinds.ILOnly, peKind);
```
We should avoid using `Assembly.LoadFrom()` in tests since that will pollute the assembly space. 

If the goal is to test that a valid assembly has been generated, would it be sufficient to use `CLRHelpers.PeVerify(string path)` instead?

am11 replied:
I haven't used CLRHelpers.PeVerify before, but it looks like net472 only. Is there a separate port for netcoreapp2?

jaredpar replied:
What is the core goal here though for `Assembly.LoadFrom`? If the goal is to inspect that the assembly is valid in some way generally we can use alternate loads that don't pollute the assembly space to verify the contents.

---
In reply to: [374409125](https://github.com/dotnet/roslyn/pull/41166#discussion_r374409125) [](ancestors = 374409125)

am11 replied:
> inspect that the assembly is valid in some way generally

Correct, to test the validity of produced assembly. I have changed it to `AssemblyName.GetAssemblyName(path)` which throws `BadImageFormatException` when assembly at `path` is invalid.

agocke replied:
I think this may work https://github.com/dotnet/roslyn/blob/2ed60699f9961a451f60d560da8bd4788c0182fc/src/Test/Utilities/Portable/Compilation/IRuntimeEnvironment.cs#L28

If you can construct a runtime environment from the emitted module, PEVerify should "just work"

am11 replied:
These are still unimplemented: https://github.com/dotnet/roslyn/blob/2ed60699f9961a451f60d560da8bd4788c0182fc/src/Test/Utilities/Portable/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs#L104-L115

so test will be a no-op on netcoreapp3.1.

cston commented on src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb:
```diff
@@ -605,6 +607,113 @@ SRC.VB(1) : error BC30037: Character is not valid.
+            If Console.IsInputRedirected Then
+                ' [applicable to both Windows and Unix]
+                ' if our parent (xunit) process itself has input redirected, we cannot test this
+                ' error case because our child process will inherit it and we cannot achive what
```
>achive [](start = 85, length = 6)

Typo

cston commented on src/Compilers/VisualBasic/Test/CommandLine/CommandLineTests.vb:
```diff
@@ -605,6 +607,113 @@ SRC.VB(1) : error BC30037: Character is not valid.
+
+            Assert.False(result.ContainsErrors, $"Compilation error(s) occurred: {result.Output} {result.Errors}")
+
+            Dim assembly As Assembly = Assembly.LoadFrom(Path.Combine(tempDir, name))
```
>Assembly.LoadFrom [](start = 39, length = 17)

Avoid `Assembly.LoadFrom()`.

cston commented:
@jaredpar, @agocke, please review, thanks.

agocke commented on src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:
```diff
@@ -5721,10 +5721,9 @@ class A                                                               \
-
-            Assert.Equal(PortableExecutableKinds.ILOnly, peKind);
+            var assemblyName = AssemblyName.GetAssemblyName(Path.Combine(tempDir, name));
+            Assert.Equal(name.Replace(".dll", ", Version=0.0.0.0, Culture=neutral, PublicKeyToken=null"),
```
Instead of testing the name, why don't you use hello world as the input source and then run the emitted exe? That's basically a full end-to-end test.

am11 replied:
This test is for compiling library (with empty method) from stdin source; the one above does the exe E2E testing.

agocke replied:
Ah, I see.

am11 commented:
I believe the CI integration [test failure](https://dev.azure.com/dnceng/public/_build/results?buildId=507187&view=ms.vss-test-web.build-test-results-tab&runId=16093238&resultId=100498&paneView=debug) is unrelated to PR changes.

agocke commented:
@am11 Agreed, the integration tests appear to have fallen over. We're working on bringing them back up

am11 commented:
@agocke, resolved the merged conflict. All green. 🌱

cston commented:
Thanks @am11 for the contribution.

am11 commented:
Thank you for a great review! 👍
